### PR TITLE
Add options for not invalidating cache when upsert

### PR DESCRIPTION
**Pull request by samUber**
Currently we invalidate the fallback cache every time we do an upset. Add options for some entities to skip invalidating.

### CODE REVIEW

File: connectors/cache/fallback.go [hunk 1]
```diff
@@ -49,12 +49,31 @@ type rangeQuery struct {
 	Limit      int
 }
 
+// Options returns a function that's being used for connector initialization
+type Options func(*Connector) error
+
+// WittSkipWriteInvalidateEntities provides the option for client to set the entites
+func WittSkipWriteInvalidateEntities(entities ...dosa.DomainObject) Options {
+	return func(c *Connector) error {
+		c.skipWriteInvalidateEntitiesMap = createSkipWriteCacheInvlidateSet(entities)
```
Comment (tiffanycheng) [inline connectors/cache/fallback.go]: spelling

File: connectors/cache/fallback.go [hunk 2]
```diff
@@ -49,12 +49,31 @@ type rangeQuery struct {
 	Limit      int
 }
 
+// Options returns a function that's being used for connector initialization
+type Options func(*Connector) error
+
+// WittSkipWriteInvalidateEntities provides the option for client to set the entites
+func WittSkipWriteInvalidateEntities(entities ...dosa.DomainObject) Options {
```
Comment (tiffanycheng) [inline connectors/cache/fallback.go]: spelling

File: connectors/cache/fallback.go [hunk 3]
```diff
@@ -393,13 +409,18 @@ func (c *Connector) cacheWrite(w func() error) error {
 	return nil
 }
 
+func (c *Connector) shouldSkipInvalidateCacheOnWrite(ei *dosa.EntityInfo) bool {
+	if c.skipWriteInvalidateEntitiesMap == nil {
```
Comment (tiffanycheng) [inline connectors/cache/fallback.go]: don't need this check, it will just return false for nil map

File: connectors/cache/fallback_test.go [hunk 4]
```diff
@@ -248,8 +258,17 @@ func TestUpsertCases(t *testing.T) {
 		},
 	}
 	for _, t := range testCases {
-		runTestCase(t)
+		runTestCaseWithInvalidateCache(t)
+	}
+	runTestCaseWithoutInvalidateCache := func(tc testCase) {
```
Comment (tiffanycheng) [inline connectors/cache/fallback_test.go]: seems like your test should be failing. the test says it's still calling Remove on the fallback

File: connectors/cache/fallback.go (hunk 3 above)
Comment (samUber) [reply connectors/cache/fallback.go]: It only returns false because of the struct initialization.

File: connectors/cache/fallback_test.go (hunk 4 above)
Comment (samUber) [reply connectors/cache/fallback_test.go]: there were some issue with the mocking originally, I will help with fixing it.

File: connectors/cache/fallback.go [hunk 5]
```diff
@@ -49,12 +49,31 @@ type rangeQuery struct {
...
+	return func(c *Connector) error {
+		c.skipWriteInvalidateEntitiesMap = createSkipWriteCacheInvalidateSet(entities)
+		return nil
+	}
+}
+
 // NewConnector creates a fallback cache connector
-func NewConnector(origin, fallback dosa.Connector, scope metrics.Scope, entities ...dosa.DomainObject) *Connector {
-	return newConnector(origin, fallback, scope, encoding.NewGobEncoder(), entities...)
+func NewConnector(origin, fallback dosa.Connector, scope metrics.Scope, entities []dosa.DomainObject, options ...Options) *Connector {
```
Comment (tiffanycheng) [inline connectors/cache/fallback.go]: i don't think this function makes much sense if you're passing in Option that can only take a fallback Connector as an argument, but this function passes in a dosa.Connector interface.

File: connectors/cache/fallback_test.go [hunk 6]
```diff
@@ -169,35 +169,66 @@ func (e *argsEncoder) Decode(byt []byte, int interface{}) error {
...
+
+	connector := NewConnector(mockOrigin, mockFallback, nil, cacheableEntities, failOption,
+		WithSkipWriteInvalidateEntities(cacheableEntities...))
+
+	assert.NotNil(t, connector)
+}
+
+// Test dosa upsert and the various behaviors of the fallback
+func TestUpsertCases(t *testing.T) {
+	runTestCaseWithInvalidateCache := func(tc testCase) {
```
Comment (tiffanycheng) [inline connectors/cache/fallback_test.go]: can we move this to line 276 to group it with the other function?

**Comment by samUber** [comment]
addressed comment

**Pull request closed without merging**